### PR TITLE
Add support for thrift exception parsing in client.

### PR DESCRIPTION
**Pull request by Raynos**
This PR adds initial thrift exception support.

 - Read thrift exceptions in the HTTP client.
 - Add support for thrif exceptions in Method class.
 - Add tests for thrift exception usage.

Next up in future PRs:

 - TODO: Support serializing thrift exceptions in server
 - TODO: Generate struct converter functions for client/server
        exceptions.
 - TODO: Ensure http endpoints convert exceptions properly.

r: @uber/zanzibar-team

### CODE REVIEW

File: codegen/templates/http_client.tmpl [hunk 1]
```diff
@@ -98,22 +99,79 @@ func (c *{{$clientName}}) {{title .Name}}(
 		respHeaders[k] = res.Header.Get(k)
 	}
 
-	res.CheckOKResponse({{.OKStatusCode.Code}})
+	res.CheckOKResponse([]int{
```
Comment (ChuntaoLu) [inline codegen/templates/http_client.tmpl]: Is this still necessary? It's covered by the default case in the switch check below

Comment (Raynos) [inline codegen/templates/http_client.tmpl]: There are 4 cases, only two of them have a switch.

File: codegen/test_data/clients/bar.gogen [hunk 2]
```diff
@@ -216,14 +303,38 @@ func (c *BarClient) TooManyArgs(
...
+
+	case 403:
+		var exception clientsBarBar.BarException
+		err = res.ReadAndUnmarshalBody(&exception)
+		if err != nil {
+			return nil, respHeaders, err
+		}
+		return nil, respHeaders, &exception
+
+	default:
```
Comment (stevededalus) [inline codegen/test_data/clients/bar.gogen]: how about removing this from default and then you don't need to worry about the dead code check below? Then anyone adding a status code but forgetting to return will see "Unexpected http client response" errors when testing/building.

File: codegen/templates/http_client.tmpl [hunk 3]
```diff
@@ -98,22 +99,79 @@ func (c *{{$clientName}}) {{title .Name}}(
...
+		default:
+			// TODO: log about unexpected body bytes?
+			_, err = res.ReadAll()
+	}
+
+	if err != nil {
+		return respHeaders, err
+	}
+	return respHeaders, nil
+	{{else if eq (len .Exceptions) 0}}
```
Comment (stevededalus) [inline codegen/templates/http_client.tmpl]: Might be slightly more readable if this was the second case in your if-else-if-else chain. That way the two switch statements are together and the two "base" cases are together.

File: codegen/test_data/clients/bar.gogen (hunk 2 above)
Comment (Raynos) [inline codegen/test_data/clients/bar.gogen]: Good suggestion.

File: codegen/templates/http_client.tmpl (hunk 1 above)
Comment (ChuntaoLu) [inline codegen/templates/http_client.tmpl]: I see, maybe it can be removed for the cases that have switch.

Comment (Raynos) [inline codegen/templates/http_client.tmpl]: This method still logs a warning to disk where as the default block only returns an error.

I want to keep the warning logic here.

Comment (zhenghuiwang) [inline codegen/templates/http_client.tmpl]: It is simpler to change `CheckOKResponse` to take `.ValidStatusCodes` as input than serialization here.

Comment (Raynos) [inline codegen/templates/http_client.tmpl]: There is no reference to .ValidStatusCodes anywhere. it only exists on EndpointSpec which is not loaded in the gateway binary only in the code generation runner.

**Pull request merged**